### ISSUE TITLE
Apartness checker should take impossible equalities into consideration

### ISSUE THREAD

**Issue by LiamGoodacre** [purescript/purescript#3329]
Given the following setup:
```purescript
class Inject f g where
  inj :: f -> g
  prj :: g -> Maybe f

instance injectRefl :: Inject x x where
  inj x = x
  prj x = Just x
else instance injectLeft :: Inject l (Either l r) where
  inj x = Left x
  prj (Left x) = Just x
  prj _ = Nothing
else instance injectRight :: Inject x r => Inject x (Either l r) where
  inj x = Right (inj x)
  prj (Right x) = prj x
  prj _ = Nothing
```

With the following scenario:
```purescript
-- should work but doesn't
injL :: forall f g. f -> Either f g
injL = inj

injR :: forall f g. g -> Either f g
injR = inj
```

In the case of `injL` the reason is that the apartness checker doesn't know that `f` can't possibly be `Either f g` because that would lead to an infinite type: `Either (Either (Either ... g) g) g`.
But `injR` shouldn't work, because we don't know that `f /~ g`.

**Comment by LiamGoodacre** [comment]
Yes that right.

**Comment by kcsongor** [comment]
Strictly speaking, this is not relevant here, because it assumes type families, but there's interesting discussion in Section 6.1 of
https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/popl137-eisenberg.pdf

TL;DR: in the presence of recursive type families, treating `f` and `Either f g` apart is unsound. Perhaps something to watch out for in the future?
(Here's the proposed solution: https://arxiv.org/pdf/1706.09715.pdf)

**Comment by LiamGoodacre** [comment]
Thanks for the resources @kcsongor!
